jenkins/check_projects.py

```diff
@@ -39,190 +41,406 @@
+            'serviceAccount:kubekins@kubernetes-jenkins.iam.gserviceaccount.com',
+            'serviceAccount:pr-kubekins@kubernetes-jenkins-pull.iam.gserviceaccount.com'],
+    },
+    'EnableVMZonalDNS': False,
 }
 
 
-def GetConfig(string):
-  if not string:
-    return DEFAULT
-  elif not os.path.isfile(string):
-    raise argparse.ArgumentTypeError('not a file: %s' % string)
-  with open(string) as fp:
-    return json.loads(fp.read())
-
-
-def CheckProjects(needed, filt, fix=False):
-  projects = LoadProjects(
-      '%s/../jobs' % os.path.dirname(__file__), filt)
-  print >>sys.stderr, 'Checking %d projects for iam bindings:' % len(projects)
-  for role, needs in sorted(needed.items()):
-    print >>sys.stderr, '  %s: %s' % (role, ','.join(Sane(n) for n in needs))
-
-  threads = [
-      threading.Thread(target=Check, args=(p, needed, fix))
-      for p in sorted(projects)
-  ]
-
-  for thread in threads:
-    thread.start()
-
-  for thread in threads:
-    thread.join()
-
-  print >>sys.stderr, '%d already configured, %d updated, %d problems' % (
-      len(completed), len(updated), len(errors))
-
-  fixers = collections.defaultdict(int)
-  unk = ['user:unknown']
-  for project in errors:
-    for name in helpers.get(project,unk):
-      fixers[name] += 1
-    print >>sys.stderr, '  %s: %s' % (
-        project, ','.join(Sane(s) for s in sorted(helpers.get(project, unk))))
-
-  print >>sys.stderr, 'Helpers:'
-  for name, count in sorted(fixers.items(), key=lambda i: i[1]):
-    print >>sys.stderr, '  %s: %s' % (count, Sane(name))
-  sys.exit(1)
-
-
-def Sane(member):
-  if ':' not in member:
-    raise ValueError(member)
-  email = member.split(':')[1]
-  return email.split('@')[0]
-
-
-def LoadProjects(configs, filt):
-  projects = set()
-  regex = re.compile(filt)
-  for dirname, _, files in os.walk(configs):
-    for path in files:
-      full_path = os.path.join(dirname, path)
-      if not regex.match(path):
-        continue
-      with open(full_path) as fp:
-        for project in re.findall('^PROJECT="?([^"\r\n]+)"?$', fp.read(), re.MULTILINE):
-          if '{' not in project:
-            projects.add(project)
-            continue
-          else:
-            raise ValueError(project)
-  return projects
-
-
-def Update(project, role, member):
-  with semaphore:
-    err = subprocess.call(
-        [
-            'gcloud', '-q', 'projects',
-            'add-iam-policy-binding',
+class RateLimitedExec(object):
+    """Runs subprocess commands with a rate limit."""
+
+    def __init__(self):
+        self.semaphore = threading.Semaphore(10)  # 10 concurrent gcloud calls
+
+    def check_output(self, *args, **kwargs):
+        with self.semaphore:
+            return subprocess.check_output(*args, **kwargs)
+
+    def call(self, *args, **kwargs):
+        with self.semaphore:
+            return subprocess.call(*args, **kwargs)
+
+
+class Results(object):
+    """Results of the check run."""
+
+    class Info(object):
```

Comment:
Consider moving the description of errors and helpers into the pydoc for Info (which lint will complain about not existing)